### PR TITLE
Unit Tests, Updated Signature, Go1.11 and Bash support, Better action-compiler and plenty of bug fix

### PR DESCRIPTION
**Pull request by sciabarracom**
The user-visible changes are:

- updated the runtime to golang 1.11
- changed the signature of actions to "func Main(args map[string]interface{}) map[string]interface{}"
- implemented all the mandatory tests for runtimes, both to GoLang and "pure" ActionLoop (with shell scripts)
- the runtime can still be used as a compiler to precompile and redure the startup time, but now in a simpler way as a filter (no more volume mounts)
- In the tests, there are also examples showing the base runtime (the action loop) is also a good runtime for shell script

Under the hood, I did a lot of work:

- internally now the compiler is a python script, no more a shell script, easier to understand an maintain
- I use now the same code to process the action at init and when I use the runtime as a compiler
- compiler is now compiling also packages and vendor folder
- I optimized code in many places to avoid converting byte in string and copying files
- I fixed a lot of bugs that emerged trying to implement the test, everything from Unicode handling to environment  variables to unexpected limits in buffers and so on
- I also refactored tests to be easier to maintain (I had since I run them a lot of times)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -158,8 +162,28 @@ As such it works with any executable that supports the following simple protocol
 
 Repeat forever:
 - read one line from standard input (file descriptor 0)
-- parse the line as a json object
-- execute the action, logging in standard output and in standard error (file descriptor 1 and 2)
+- parse the line as a json object, that will be in format:
+
+```{
```
Comment (csantanapr) [inline README.md]: I think the contract states that `value` will be the one present, and any other fields in the JSON will be converted to environment variables.

File: README.md [hunk 2]
```diff
@@ -158,8 +162,28 @@ As such it works with any executable that supports the following simple protocol
...
+ "namespace": String,
+ "action_name": String,
+ "api_host": String,
+ "api_key": String,
+ "activation_id": String,
+ "deadline": Number
+}```
+
+Note that if you use libraries, those will expect the values in enviroment variables:
+
```
Comment (csantanapr) [inline README.md]: Same comment here, this set is a base set of environment variables but more environment variables should be allow to be passed as additional fields.

File: common/gobuild.py.launcher.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	// assign the main function
+	type Action func(event map[string]interface{}) map[string]interface{}
+	var action Action
+	action = Main
+
+	// Params are the parameters sent to the action
+	type Params struct {
+		Value        *map[string]interface{} `json:"value"`
+		Namespace    string                  `json:"namespace"`
```
Comment (csantanapr) [inline common/gobuild.py.launcher.go]: We should not hardcode the set of environment variables, other operator like public cloud multi-tenant would passing more environment variables, so the parsing of the json input should to take `value` as input params, and any other fields present if any treat them as environment variables by doing the upper case and appending the `__OW_` prefix.

Comment (csantanapr) [reply common/gobuild.py.launcher.go]: Here are some examples on how to process the `/run` input in other runtimes:
https://github.com/apache/incubator-openwhisk-runtime-nodejs/blob/master/core/nodejsActionBase/src/service.js#L160
https://github.com/apache/incubator-openwhisk-runtime-java/blob/master/core/java8/proxy/src/main/java/openwhisk/java/action/Proxy.java#L148

File: README.md (hunk 1 above)
Comment (sciabarracom) [reply README.md]: Ok, will fix it.

File: common/gobuild.py.launcher.go (hunk 3 above)
Comment (sciabarracom) [reply common/gobuild.py.launcher.go]: Agreed

**Comment by sciabarracom** [comment]
Volume problems: inconsistent behavior between Mac and Linux. Using docker -v it was working on Mac and giving problems in Linux: for example if the output directory was already created, it was not writable. If I did not create the directory, it was not created on Mac.   Also, the compilation script had to work in different situations, requiring paths to be created. I ended up to keep all the compilation inside the runtime and add an option to the proxy to read the input, use the same code that handles the "/init" and then dumping the result in the stdout. It was cleaner and simpler.

**Comment by sciabarracom** [comment]
not sure what is causing this conflict however you can remote this file it is irrelevant

**Pull request merged**